File: native/ios/SecureKeypadPresentationContractTest.swift
import Foundation

private struct ContractRandomGenerator: RandomNumberGenerator {
    private var state: UInt64 = 0x9e3779b97f4a7c15

    mutating func next() -> UInt64 {
        state = state &* 6_364_136_223_846_793_005 &+ 1
        return state
    }
}

@main
struct SecureKeypadPresentationContractTest {
    static func main() {
        precondition(!secureKeypadShouldProtectPresentation(applicationIsActive: true, screenIsCaptured: false))
        precondition(secureKeypadShouldProtectPresentation(applicationIsActive: false, screenIsCaptured: false))
        precondition(secureKeypadShouldProtectPresentation(applicationIsActive: true, screenIsCaptured: true))
        precondition(secureKeypadMaskedDisplayText(length: 0, protected: false) == "")
        precondition(secureKeypadMaskedDisplayText(length: 3, protected: false) == "•••")
        precondition(secureKeypadMaskedDisplayText(length: 3, protected: true) == "Protected")
        precondition(secureKeypadAccessibilityLabel(length: 0, protected: false) == "No input")
        precondition(secureKeypadAccessibilityLabel(length: 3, protected: false) == "3 characters entered")
        precondition(secureKeypadAccessibilityLabel(length: 3, protected: true) == "Protected")
        precondition(!secureKeypadIsValidRenderedLength(-1))
        precondition(!secureKeypadIsValidRenderedLength(4_097))
        precondition(secureKeypadIsValidDisplayState(0))
        precondition(secureKeypadIsValidDisplayState(3))
        precondition(!secureKeypadIsValidDisplayState(4))
        precondition(secureKeypadDisplayStateName(0) == "empty")
        precondition(secureKeypadDisplayStateName(3) == "cancelled")
        precondition(secureKeypadDisplayStateName(4) == "invalid")
        let activeSnapshot = [
            "displayState=masked",
            "maskedDisplay=•••",
            "accessibility=3 characters entered",
            "protected=false",
        ].joined(separator: "\n")
        precondition(secureKeypadSecuritySnapshot(
            length: 3,
            protected: false,
            displayState: 1
        ) == activeSnapshot)
        let protectedSnapshot = [
            "displayState=masked",
            "maskedDisplay=Protected",
            "accessibility=Protected",
            "protected=true",
        ].joined(separator: "\n")
        precondition(secureKeypadSecuritySnapshot(
            length: 3,
            protected: true,
            displayState: 1
        ) == protectedSnapshot)
        precondition(secureKeypadSecuritySnapshot(length: 3, protected: false, displayState: 4) == nil)
        precondition(secureKeypadMonotonicCommandDecision(previous: nil, requestId: 0) == .accept)
        precondition(secureKeypadMonotonicCommandDecision(previous: 4, requestId: 5) == .accept)
        precondition(secureKeypadMonotonicCommandDecision(previous: 4, requestId: 4) == .ignore)
        precondition(secureKeypadMonotonicCommandDecision(previous: 4, requestId: 3) == .invalid)

        let rows = [
            [(id: "digit-1", input: true), (id: "backspace", input: false), (id: "digit-2", input: true)],
            [(id: "digit-3", input: true), (id: "submit", input: false), (id: "digit-4", input: true)],
        ]
        var generator = ContractRandomGenerator()
        let randomized = secureKeypadPresentationRows(
            rows,
            randomizeInputKeys: true,
            isInput: { $0.input },
            using: &generator
        )
        precondition(randomized.map(\.count) == rows.map(\.count))
        precondition(randomized[0][1].id == "backspace")
        precondition(randomized[1][1].id == "submit")
        precondition(Set(randomized.flatMap { $0 }.filter { $0.input }.map(\.id)) == Set([
            "digit-1", "digit-2", "digit-3", "digit-4",
        ]))
        let unchanged = secureKeypadPresentationRows(
            rows,
            randomizeInputKeys: false,
            isInput: { $0.input },
            using: &generator
        )
        precondition(unchanged.map { $0.map(\.id) } == rows.map { $0.map(\.id) })
    }
}

File: native/android/SecureKeypadPresentationContractTest.kt
package com.uulab.securekeypad

/** Standalone accessibility/masking contract test; no Android runtime is required. */
fun main() {
    check(secureKeypadMaskedDisplayText(0) == "")
    check(secureKeypadMaskedDisplayText(3) == "•••")
    check(secureKeypadAccessibilityLabel(0) == "No input")
    check(secureKeypadAccessibilityLabel(3) == "3 characters entered")
    check(secureKeypadMaskedDisplayText(3, protected = true) == "Protected")
    check(secureKeypadAccessibilityLabel(3, protected = true) == "Protected")
    check(runCatching { secureKeypadMaskedDisplayText(-1) }.isFailure)
    check(runCatching { secureKeypadAccessibilityLabel(4_097) }.isFailure)
    check(secureKeypadIsValidDisplayState(0))
    check(secureKeypadIsValidDisplayState(3))
    check(!secureKeypadIsValidDisplayState(4))
    check(secureKeypadDisplayStateName(0) == "empty")
    check(secureKeypadDisplayStateName(3) == "cancelled")
    check(secureKeypadDisplayStateName(4) == "invalid")
    val activeSnapshot = listOf(
        "displayState=masked",
        "maskedDisplay=•••",
        "accessibility=3 characters entered",
        "protected=false",
    ).joinToString("\n")
    check(secureKeypadSecuritySnapshot(length = 3, protected = false, displayState = 1) == activeSnapshot)
    val protectedSnapshot = listOf(
        "displayState=masked",
        "maskedDisplay=Protected",
        "accessibility=Protected",
        "protected=true",
    ).joinToString("\n")
    check(secureKeypadSecuritySnapshot(length = 3, protected = true, displayState = 1) == protectedSnapshot)
    check(runCatching {
        secureKeypadSecuritySnapshot(length = 3, protected = false, displayState = 4)
    }.isFailure)
    check(secureKeypadDecodeMaskedState(Long.MIN_VALUE) == null)
    check(secureKeypadDecodeMaskedState((3L shl 32) or 1L) == (3 to 1))
    check(secureKeypadMonotonicCommandDecision(null, 0) == SecureKeypadCommandDecision.ACCEPT)
    check(secureKeypadMonotonicCommandDecision(4, 5) == SecureKeypadCommandDecision.ACCEPT)
    check(secureKeypadMonotonicCommandDecision(4, 4) == SecureKeypadCommandDecision.IGNORE)
    check(secureKeypadMonotonicCommandDecision(4, 3) == SecureKeypadCommandDecision.INVALID)
}
